ethcore/private-tx/src/lib.rs

```diff
@@ -567,105 +631,44 @@ impl Importer for Arc<Provider> {
 			.iter()
 			.find(|address| self.validator_accounts.contains(address));
 
-		match validation_account {
-			None => {
-				// TODO [ToDr] This still seems a bit invalid, imho we should still import the transaction to the pool.
-				// Importing to pool verifies correctness and nonce; here we are just blindly forwarding.
-				//
-				// Not for verification, broadcast further to peers
-				self.broadcast_private_transaction(rlp.into());
-				return Ok(());
-			},
-			Some(&validation_account) => {
-				let hash = private_tx.hash();
-				trace!("Private transaction taken for verification");
-				let original_tx = self.extract_original_transaction(private_tx, &contract)?;
-				trace!("Validating transaction: {:?}", original_tx);
-				// Verify with the first account available
-				trace!("The following account will be used for verification: {:?}", validation_account);
-				let nonce_cache = Default::default();
-				self.transactions_for_verification.lock().add_transaction(
-					original_tx,
-					contract,
-					validation_account,
-					hash,
-					self.pool_client(&nonce_cache),
-				)?;
-				let provider = Arc::downgrade(self);
-				self.channel.send(ClientIoMessage::execute(move |_| {
+		let original_tx = self.extract_original_transaction(private_tx.clone(), &contract)?;
+		trace!("Original transaction: {:?}", original_tx);
+		let nonce_cache = Default::default();
+		self.transactions_for_verification.add_transaction(
+			original_tx,
+			validation_account.map(|&account| account),
+			private_tx.clone(),
+			self.pool_client(&nonce_cache),
+		)?;
+		let provider = Arc::downgrade(self);
+		if let Err(e) = self.channel.send(ClientIoMessage::execute(move |_| {
 					if let Some(provider) = provider.upgrade() {
-						if let Err(e) = provider.process_queue() {
+						if let Err(e) = provider.process_verification_queue() {
```

Comment:
Same as with `process_signatures` might be easier to just offload everything inside the `IoWorker` thread and just `move` the `private_tx` here and then import and process it. Currently we are almost always processing only one transaction, but the whole code assumes we are operating on lists.